### PR TITLE
Pipeline & Terraform changes to deploy to staging & production

### PR DESCRIPTION
**Pull request by vigneshmsft**
### Context
Config to deploy to staging and production in PaaS

### Changes proposed in this pull request
Move application environment variables to a YAML variable (stored in GitHub secrets for now).
Read and apply above yaml config using terraform as app environment variables.
Change deploy pipeline to use correct secret value depending on `environment` being deployed.

Example deployments :

qa                : https://github.com/DFE-Digital/find-teacher-training/runs/1285870337
staging        : https://github.com/DFE-Digital/find-teacher-training/actions/runs/318283816
production  : https://github.com/DFE-Digital/find-teacher-training/actions/runs/318284149

### Trello card
https://trello.com/c/A4528lno/2382-deploy-to-staging-and-production-in-paas

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally
- [x] Product review

### CODE REVIEW

File: terraform/modules/paas/variables.tf [hunk 1]
```diff
@@ -16,41 +16,28 @@ variable web_app_memory { default = 512 }
...
-    SECRET_KEY_BASE                = var.secret_key_base
-    SENTRY_DSN                     = var.sentry_dsn
-    SETTINGS__CYCLE_ENDING_SOON    = var.settings_cycle_ending_soon
-    SETTINGS__CYCLE_HAS_ENDED      = var.settings_cycle_has_ended
-    SETTINGS__DISPLAY_APPLY_BUTTON = var.settings_display_apply_button
-    SETTINGS__GOOGLE__GCP_API_KEY  = var.settings_google_gcp_api_key
-    SETTINGS__GOOGLE__MAPS_API_KEY = var.settings_google_maps_api_key
+  common_app_config = {
+    ASSETS_PRECOMPILE         = true
+    AUTHORISED_HOSTS          = "${local.web_app_name}.london.cloudapps.digital"
```
Comment (saliceti) [inline terraform/modules/paas/variables.tf:34]: Can you try to remove it if it's not used by rails?

File: terraform/variables.tf [hunk 2]
```diff
@@ -14,22 +14,21 @@ variable paas_web_app_memory { default = 512 }
...
-variable paas_settings_google_maps_api_key {}
-
 variable paas_logstash_url {}
 
+variable paas_app_config_file { default = "./app_config.yml" }
+
 locals {
   cf_api_url = "https://api.london.cloud.service.gov.uk"
+  paas_app_config = merge(yamldecode(file(var.paas_app_config_file)), {
+    SETTINGS__CYCLE_ENDING_SOON    = var.paas_settings_cycle_ending_soon
```
Comment (saliceti) [inline terraform/variables.tf:30]: Can we move this to separate yml files per environment stored in the git repo?

File: terraform/variables.tf [hunk 3]
```diff
@@ -14,22 +14,21 @@ variable paas_web_app_memory { default = 512 }
...
 
 variable paas_settings_display_apply_button {}
 
-variable paas_settings_google_gcp_api_key {}
-
-variable paas_settings_google_maps_api_key {}
-
 variable paas_logstash_url {}
 
+variable paas_app_config_file { default = "./app_config.yml" }
```
Comment (saliceti) [inline terraform/variables.tf:25]: Can we rename it to something like app_secrets to make it more explicit?

File: terraform/variables.tf [hunk 4]
```diff
@@ -12,24 +12,15 @@ variable paas_web_app_instances { default = 1 }
...
-variable paas_settings_google_maps_api_key {}
+variable paas_app_config_file { default = "./workspace_variables/app_config.yml" }
 
-variable paas_logstash_url {}
+variable paas_app_secrets_file { default = "./app_secrets.yml" }
 
 locals {
-  cf_api_url = "https://api.london.cloud.service.gov.uk"
+  cf_api_url                     = "https://api.london.cloud.service.gov.uk"
+  app_config                     = yamldecode(file(var.paas_app_config_file))
```
Comment (saliceti) [inline terraform/variables.tf:23]: Isn't it `yamldecode(file(var.paas_app_config_file))[var.paas_app_environment]` ?

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -25,27 +25,24 @@ jobs:
...
         env:
-          ARM_ACCESS_KEY: ${{ secrets.TERRAFORM_STATE_ACCESS_KEY }}
-      
-      - name: Terraform plan & apply
+          APP_SECRETS: ${{ secrets[format('APP_SECRETS_{0}', env.DEPLOY_ENV)] }}
+
+      - name: Terraform init, plan & apply
         id: terraform_deploy
         run: |
+          terraform init -backend-config=workspace_variables/${{ env.DEPLOY_ENV }}_backend.tfvars
```
Comment (saliceti) [inline .github/workflows/deploy.yml:38]: You can use `${DEPLOY_ENV}` as the variable is available in the shell environment

File: terraform/variables.tf (hunk 4 above)
Comment (vigneshmsft) [reply terraform/variables.tf]: yes, I must have reversed the two vars, fixed it now. thanks

**Review by saliceti — APPROVED** [review]

**Pull request merged**